// @apollo/space-kit/Table/Table.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _tslib = require('../_tslib-bcbe0269.js');
var core = require('@emotion/core');
require('../fonts/fonts.js');
var base$1 = require('../base-1dde5d6e.js');
require('../mono-f4fb9dce.js');
var colors = require('../colors/colors.js');

/**
 * Tables provide a structure to data and a visual grid making it easier to see
 * relationships and are one of the most useful tools and formats for organizing
 * and communiting structured data.
 *
 * @see https://zpl.io/bAlrjJe
 */
function Table(_a) {
    var data = _a.data, _b = _a.density, density = _b === void 0 ? "standard" : _b, columns = _a.columns, keyOn = _a.keyOn;
    var padding = density === "standard" ? 8 : density === "condensed" ? 3 : 11;
    var getRowKey = typeof keyOn === "function" ? keyOn : function (row) { return row[keyOn]; };
    return (core.jsx("table", { css: {
            borderCollapse: "collapse",
            width: "100%",
        } },
        core.jsx("colgroup", null, columns.map(function (_a) {
            var colProps = _a.colProps, id = _a.id;
            return (core.jsx("col", _tslib.__assign({ key: id }, colProps)));
        })),
        core.jsx("thead", null,
            core.jsx("tr", { css: {
                    borderBottom: "1px solid " + colors.colors.silver.dark,
                } }, columns.map(function (_a, colIndex) {
                var headerTitle = _a.headerTitle, id = _a.id;
                return (core.jsx("th", { key: id, css: _tslib.__assign(_tslib.__assign({}, base$1.xsmall), { textTransform: "uppercase", color: colors.colors.grey.darker, fontWeight: 600, textAlign: "left", padding: padding, paddingLeft: colIndex === 0 ? 0 : padding, paddingRight: colIndex === columns.length - 1 ? 0 : padding }) }, headerTitle));
            }))),
        core.jsx("tbody", null, data.map(function (item, index) { return (core.jsx("tr", { key: getRowKey(item) }, columns.map(function (_a, colIndex) {
            var render = _a.render, id = _a.id;
            return (core.jsx("td", { key: id, css: {
                    // no border on the bottom row
                    borderBottom: index === data.length - 1
                        ? "none"
                        : "1px solid " + colors.colors.silver.dark,
                    padding: padding,
                    paddingLeft: colIndex === 0 ? 0 : padding,
                    paddingRight: colIndex === columns.length - 1 ? 0 : padding,
                } }, render(item, index, data)));
        }))); }))));
}

exports.Table = Table;
