// gatsby-interface/dist/src/components/DropdownMenu/DropdownMenu.styles.esm.js
import { taggedTemplateLiteral as _taggedTemplateLiteral, defineProperty as _defineProperty } from '../../../_virtual/_rollupPluginBabelHelpers.js';
import { keyframes } from '@emotion/core';

function _templateObject() {
  var data = _taggedTemplateLiteral(["\nto {\n  transform: translate(0, 0) perspective(1000px) rotateX(0);\n}\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}
var enter = keyframes(_templateObject());
var dropdownCss = function dropdownCss(theme) {
  return _defineProperty({
    background: theme.colors.primaryBackground,
    border: "1px solid ".concat(theme.colors.grey[10]),
    borderRadius: theme.radii[3],
    outline: "none",
    boxShadow: theme.shadows.floating,
    width: "80vw",
    transform: "translate(0, 0) perspective(1000px) rotateX(-35deg)",
    animation: "".concat(enter, " 0.5s ease forwards"),
    transformOrigin: "top center",
    overflow: "auto",
    maxHeight: "20rem",
    "[data-selected]": {
      background: theme.colors.purple[10],
      color: theme.colors.gatsby
    }
  }, theme.mediaQueries.phablet, {
    maxWidth: "20rem"
  });
};
var menuItemCss = function menuItemCss(theme) {
  return {
    cursor: "pointer",
    color: theme.colors.grey[90],
    fontSize: theme.fontSizes[1],
    padding: "".concat(theme.space[4], " ").concat(theme.space[5]),
    lineHeight: theme.lineHeights["default"],
    overflow: "hidden",
    wordWrap: "break-word",
    overflowWrap: "break-word",
    hyphens: "auto",
    display: "block",
    textDecoration: "none"
  };
};
var dropdownButtonCss = function dropdownButtonCss(theme) {
  return {
    cursor: "pointer",
    display: "flex",
    alignItems: "center",
    border: "1px solid ".concat(theme.colors.grey[30]),
    borderRadius: theme.radii[2],
    color: theme.colors.grey[90],
    fontSize: theme.fontSizes[2],
    padding: "".concat(theme.space[2], " ").concat(theme.space[4]),
    width: "100%",
    "&:focus": {
      borderColor: theme.colors.purple[40],
      boxShadow: "0 0 0 3px ".concat(theme.colors.purple[20]),
      outline: 0,
      transition: "box-shadow 0.15s ease-in-out"
    }
  };
};
var dropdownButtonIconCss = function dropdownButtonIconCss(_theme) {
  return {
    marginInlineStart: "auto",
    transition: "transform .3s",
    "[aria-expanded] > &": {
      transform: "rotate(180deg)"
    }
  };
};

export { dropdownButtonCss, dropdownButtonIconCss, dropdownCss, menuItemCss };
