// gatsby-interface/dist/src/components/form/components/styled-primitives/StyledTextArea.esm.js
import { extends as _extends } from '../../../../../_virtual/_rollupPluginBabelHelpers.js';
import { jsx } from '@emotion/core';
import React__default from 'react';
import { baseInputCss } from '../../styles/index.esm.js';

var StyledTextArea = React__default.forwardRef(function StyledTextArea(props, ref) {
  var placeholder = props.placeholder && props.disabled ? "The field is disabled" : props.placeholder;

  var baseCss = function baseCss(theme) {
    return [baseInputCss(theme), {
      display: "block",
      minHeight: "4.85em",
      resize: "vertical",
      padding: "".concat(theme.space[2], " ").concat(theme.space[3])
    }];
  };

  return jsx("textarea", _extends({
    ref: ref,
    css: baseCss
  }, props, {
    placeholder: placeholder
  }));
});
StyledTextArea.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "StyledTextArea"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form/components/styled-primitives/StyledTextArea.tsx"] = {
    name: "StyledTextArea",
    docgenInfo: StyledTextArea.__docgenInfo,
    path: "src/components/form/components/styled-primitives/StyledTextArea.tsx"
  };
}

export { StyledTextArea };
