// gatsby-interface/dist/src/components/Chip/BaseChip.esm.js
import { extends as _extends } from '../../../_virtual/_rollupPluginBabelHelpers.js';
import { jsx, css } from '@emotion/core';
import { fontSizes as b } from '../../../node_modules/gatsby-design-tokens/dist/index.esm.esm.js';
import colors from '../../theme/colors.esm.js';

var SIZE = "1.75rem";
var BORDER_WIDTH = "1px";
var BACKGROUND_COLOR = colors.grey["10"];
var baseCss = css({
  display: "inline-flex",
  alignItems: "center",
  verticalAlign: "middle",
  minWidth: SIZE,
  height: SIZE,
  padding: "0 0.75rem",
  backgroundColor: BACKGROUND_COLOR,
  borderRadius: "0.75rem",
  fontSize: b[0],
  lineHeight: "calc(".concat(SIZE, " - 2 * ").concat(BORDER_WIDTH, ")"),
  color: colors.blackFade["70"],
  border: "".concat(BORDER_WIDTH, " solid ").concat(BACKGROUND_COLOR),
  whiteSpace: "nowrap"
});
function BaseChip(_ref) {
  var props = _extends({}, _ref);

  return jsx("span", _extends({
    css: baseCss
  }, props));
}
BaseChip.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "BaseChip"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/Chip/BaseChip.tsx"] = {
    name: "BaseChip",
    docgenInfo: BaseChip.__docgenInfo,
    path: "src/components/Chip/BaseChip.tsx"
  };
}

export default BaseChip;
