// gatsby-interface/dist/src/components/StepIndicator/StepIndicatorStep.styles.esm.js
import { defineProperty as _defineProperty } from '../../../_virtual/_rollupPluginBabelHelpers.js';
import { visuallyHiddenCss, resetVisuallyHiddenCss } from '../../stylesheets/a11y.esm.js';

var baseCss = function baseCss(theme) {
  return _defineProperty({
    display: "inline-flex",
    alignItems: "center",
    borderRadius: "".concat(theme.radii[2], " ").concat(theme.radii[2], " 0 0"),
    backgroundColor: theme.colors.secondaryBackground,
    border: "1px solid ".concat(theme.colors.standardLine),
    color: theme.colors.blackFade[70],
    fontFamily: theme.fonts.heading,
    fontSize: theme.fontSizes[3],
    fontWeight: theme.fontWeights.bold,
    height: "3.5rem",
    margin: "0 ".concat(theme.space[2]),
    padding: "".concat(theme.space[1], " ").concat(theme.space[5], " 0"),
    position: "relative",
    zIndex: 0
  }, theme.mediaQueries.phablet, {
    padding: "".concat(theme.space[1], " ").concat(theme.space[7], " 0")
  });
};
var activeCss = function activeCss(theme) {
  return {
    background: theme.colors.primaryBackground,
    borderBottomColor: theme.colors.secondaryBackground,
    color: theme.colors.grey[90],
    zIndex: 2,
    ":before, :after": {
      border: "1px solid ".concat(theme.colors.standardLine),
      bottom: "-1px",
      content: "\"\"",
      height: "6px",
      position: "absolute",
      width: "6px"
    },
    ":before": {
      borderBottomRightRadius: "6px",
      borderWidth: "0 1px 1px 0",
      boxShadow: "2px 2px 0 #fff",
      left: "-6px"
    },
    ":after": {
      borderBottomLeftRadius: "6px",
      borderWidth: "0 0 1px 1px",
      boxShadow: "-2px 2px 0 #fff",
      right: "-6px"
    }
  };
};
var linkCss = function linkCss(theme) {
  return {
    display: "flex",
    alignItems: "center",
    textDecoration: "none",
    color: theme.colors.purple[50],
    "&:active": {
      color: "inherit"
    }
  };
};
var labelCss = function labelCss(_theme) {
  return {};
};
var labelInactiveCss = function labelInactiveCss(theme) {
  return [visuallyHiddenCss, _defineProperty({}, theme.mediaQueries.tablet, resetVisuallyHiddenCss)];
};
var labelActiveCss = function labelActiveCss(_theme) {
  return {};
};
var stepNumberCss = function stepNumberCss(theme) {
  var _ref3;

  return _ref3 = {
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    flexShrink: 0,
    background: theme.colors.grey[40],
    color: theme.colors.white,
    fontFamily: theme.fonts.system,
    fontSize: theme.fontSizes[0],
    fontWeight: theme.fontWeights.bold,
    lineHeight: theme.lineHeights.solid,
    height: "20px",
    width: "20px",
    borderRadius: theme.radii[6]
  }, _defineProperty(_ref3, theme.mediaQueries.tablet, {
    marginRight: theme.space[4]
  }), _defineProperty(_ref3, ":before", {
    counterIncrement: "section",
    content: "counters(section,\".\") \" \""
  }), _ref3;
};
var stepNumberActiveCss = function stepNumberActiveCss(theme) {
  return {
    background: theme.colors.purple[40],
    marginRight: theme.space[4]
  };
};
var stepNumberDoneCss = function stepNumberDoneCss(theme) {
  return {
    background: theme.colors.purple[30]
  };
};

export { activeCss, baseCss, labelActiveCss, labelCss, labelInactiveCss, linkCss, stepNumberActiveCss, stepNumberCss, stepNumberDoneCss };
