// gatsby-interface/dist/src/components/StepIndicator/StepIndicator.esm.js
import { jsx } from '@emotion/core';
import _pt from 'prop-types';

var baseCss = function baseCss(theme) {
  return {
    listStyle: "none",
    margin: 0,
    padding: 0,
    position: "relative",
    display: "flex",
    justifyContent: "center",
    textAlign: "center",
    paddingTop: theme.space[9],
    counterReset: "section",
    ":after": {
      borderBottom: "1px solid ".concat(theme.colors.standardLine),
      bottom: 0,
      content: "\"\"",
      left: 0,
      position: "absolute",
      width: "100%",
      zIndex: 1
    }
  };
};

function StepIndicator(_ref) {
  var children = _ref.children;
  return jsx("ul", {
    css: function css(theme) {
      return [baseCss(theme)];
    }
  }, children);
}
StepIndicator.propTypes = {
  children: _pt.node
};
StepIndicator.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "StepIndicator",
  "props": {
    "children": {
      "required": false,
      "tsType": {
        "name": "ReactReactNode",
        "raw": "React.ReactNode"
      },
      "description": ""
    }
  }
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/StepIndicator/StepIndicator.tsx"] = {
    name: "StepIndicator",
    docgenInfo: StepIndicator.__docgenInfo,
    path: "src/components/StepIndicator/StepIndicator.tsx"
  };
}

export { StepIndicator };
