// gatsby-interface/dist/src/components/Toast/Toast.esm.js
import { taggedTemplateLiteral as _taggedTemplateLiteral } from '../../../_virtual/_rollupPluginBabelHelpers.js';
import { keyframes, css, jsx } from '@emotion/core';
import _pt from 'prop-types';
import { MdClose, MdDone, MdWarning } from 'react-icons/md';
import { radii as S, fontSizes as b } from '../../../node_modules/gatsby-design-tokens/dist/index.esm.esm.js';
import colors from '../../theme/colors.esm.js';
import space from '../../theme/space.esm.js';
import Alert from '../../../node_modules/@reach/alert/dist/alert.esm.esm.js';
import dimensions from '../../theme/dimensions.esm.js';

function _templateObject5() {
  var data = _taggedTemplateLiteral(["\n  align-items: center;\n  background: none;\n  border: none;\n  color: ", ";\n  cursor: pointer;\n  display: flex;\n  height: ", ";\n  justify-content: center;\n  width: ", ";\n"]);

  _templateObject5 = function _templateObject5() {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = _taggedTemplateLiteral(["\n  align-items: center;\n  color: ", ";\n  display: flex;\n"]);

  _templateObject4 = function _templateObject4() {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = _taggedTemplateLiteral(["\n  line-height: 1;\n  margin: 0 ", " 0 ", ";\n"]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = _taggedTemplateLiteral(["\n  align-items: center;\n  animation: ", " 0.5s 0.25s ease forwards;\n  background: ", ";\n  border-left: 8px solid ", ";\n  border-radius: ", " ", " 0 0;\n  color: ", ";\n  display: flex;\n  font-size: ", ";\n  min-height: ", ";\n  max-width: calc(100% - (", " * 2));\n  padding-left: ", ";\n  transform: perspective(1000px) rotateX(90deg);\n  transform-origin: bottom center;\n\n  svg {\n    height: auto;\n    width: calc(", " * 0.4);\n  }\n\n  &:not(:first-of-type) {\n    border-radius: ", ";\n    margin-bottom: ", ";\n  }\n"]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = _taggedTemplateLiteral(["\n  100% {\n     transform: perspective(1000px) rotateX(0);\n  }\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}
var toastEntryAnimation = keyframes(_templateObject());
var toastCss = css(_templateObject2(), toastEntryAnimation, colors.grey[90], colors.green[50], S[2], S[2], colors.green[5], b[1], dimensions.toast.minHeight, space[7], space[4], dimensions.toast.minHeight, S[2], space[1]);
var messageCss = css(_templateObject3(), space[2], space[3]);
var statusCss = css(_templateObject4(), colors.green[50]);
var closeButtonCss = css(_templateObject5(), colors.grey[40], dimensions.toast.minHeight, dimensions.toast.minHeight);
var toastColorByTone = {
  SUCCESS: colors.green[50],
  DANGER: colors.red[60]
};
var ToastIconByTone = {
  SUCCESS: MdDone,
  DANGER: MdWarning
};
var Toast = function Toast(_ref) {
  var message = _ref.message,
      tone = _ref.tone,
      closeButtonLabel = _ref.closeButtonLabel,
      onClose = _ref.onClose;
  var IconComponent = ToastIconByTone[tone];
  return jsx(Alert, {
    css: [toastCss, css({
      borderLeftColor: toastColorByTone[tone]
    })],
    "data-testid": "toast",
    type: tone === "DANGER" ? "assertive" : "polite"
  }, jsx("span", {
    css: [statusCss, css({
      color: toastColorByTone[tone]
    })]
  }, jsx(IconComponent, null)), jsx("div", {
    css: messageCss
  }, message), jsx("button", {
    css: closeButtonCss,
    type: "button",
    onClick: onClose,
    "aria-label": closeButtonLabel
  }, jsx(MdClose, null)));
};
Toast.propTypes = {
  message: _pt.node.isRequired,
  onClose: _pt.func.isRequired,
  closeButtonLabel: _pt.string.isRequired
};
Toast.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "Toast"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/Toast/Toast.tsx"] = {
    name: "Toast",
    docgenInfo: Toast.__docgenInfo,
    path: "src/components/Toast/Toast.tsx"
  };
}

export { Toast };
