// gatsby-interface/dist/src/components/form/components/InputField.esm.js
import { extends as _extends, objectWithoutProperties as _objectWithoutProperties } from '../../../../_virtual/_rollupPluginBabelHelpers.js';
import { jsx } from '@emotion/core';
import React__default from 'react';
import { showFormFieldDeprecatedMessage } from '../../../utils/maintenance/deprecationMessages.esm.js';
import { getInputStyles } from './FormField.helpers.esm.js';
import { useStyledFieldLabel, useStyledFieldError, useStyledFieldHint } from './FormField.esm.js';
import { useFormFieldSkeleton } from '../../form-skeletons/components/FormFieldSkeleton.esm.js';
import { InputFieldSkeleton, InputFieldSkeletonControl, InputFieldSkeletonLabel, InputFieldSkeletonError, InputFieldSkeletonHint } from '../../form-skeletons/components/InputFieldSkeleton.esm.js';

function InputField(props) {
  showFormFieldDeprecatedMessage("InputField");
  return jsx(InputFieldSkeleton, props);
}
var InputFieldControl = React__default.forwardRef(function InputFieldControl(props, ref) {
  var _useFormFieldSkeleton = useFormFieldSkeleton(),
      hasError = _useFormFieldSkeleton.hasError;

  var placeholder = props.placeholder && props.disabled ? "The field is disabled" : props.placeholder;
  return jsx(InputFieldSkeletonControl, _extends({
    ref: ref,
    css: function css(theme) {
      return [getInputStyles(theme, hasError)];
    }
  }, props, {
    placeholder: placeholder
  }));
});
function InputFieldLabel(_ref) {
  var children = _ref.children,
      size = _ref.size,
      isRequired = _ref.isRequired,
      props = _objectWithoutProperties(_ref, ["children", "size", "isRequired"]);

  var styledProps = useStyledFieldLabel(children, {
    size: size,
    isRequired: isRequired
  });
  return jsx(InputFieldSkeletonLabel, _extends({}, props, styledProps));
}
function InputFieldError(_ref2) {
  var children = _ref2.children,
      props = _objectWithoutProperties(_ref2, ["children"]);

  var styledProps = useStyledFieldError(children);
  return jsx(InputFieldSkeletonError, _extends({}, props, styledProps));
}
function InputFieldHint(props) {
  var styledProps = useStyledFieldHint();
  return jsx(InputFieldSkeletonHint, _extends({}, props, styledProps));
}
InputField.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "InputField"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form/components/InputField.tsx"] = {
    name: "InputField",
    docgenInfo: InputField.__docgenInfo,
    path: "src/components/form/components/InputField.tsx"
  };
}

InputFieldControl.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "InputFieldControl"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form/components/InputField.tsx"] = {
    name: "InputFieldControl",
    docgenInfo: InputFieldControl.__docgenInfo,
    path: "src/components/form/components/InputField.tsx"
  };
}

InputFieldLabel.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "InputFieldLabel"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form/components/InputField.tsx"] = {
    name: "InputFieldLabel",
    docgenInfo: InputFieldLabel.__docgenInfo,
    path: "src/components/form/components/InputField.tsx"
  };
}

InputFieldError.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "InputFieldError"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form/components/InputField.tsx"] = {
    name: "InputFieldError",
    docgenInfo: InputFieldError.__docgenInfo,
    path: "src/components/form/components/InputField.tsx"
  };
}

InputFieldHint.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "InputFieldHint"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form/components/InputField.tsx"] = {
    name: "InputFieldHint",
    docgenInfo: InputFieldHint.__docgenInfo,
    path: "src/components/form/components/InputField.tsx"
  };
}

export { InputField, InputFieldControl, InputFieldError, InputFieldHint, InputFieldLabel };
