// gatsby-interface/dist/src/components/form/components/InputConnectedField.esm.js
import { slicedToArray as _slicedToArray, extends as _extends } from '../../../../_virtual/_rollupPluginBabelHelpers.js';
import { jsx } from '@emotion/core';
import React__default from 'react';
import _pt from 'prop-types';
import { InputFieldBlock } from './InputFieldBlock.esm.js';
import { useConnectedField } from '../hooks/useConnectedField.esm.js';

var InputConnectedField = React__default.forwardRef(function InputConnectedField(props, ref) {
  var _useConnectedField = useConnectedField(props.name),
      _useConnectedField2 = _slicedToArray(_useConnectedField, 1),
      connectedProps = _useConnectedField2[0];

  return jsx(InputFieldBlock, _extends({
    ref: ref
  }, connectedProps, props));
});
InputConnectedField.propTypes = {
  name: _pt.string.isRequired,
  id: _pt.string,
  label: _pt.node
};
InputConnectedField.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "InputConnectedField"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form/components/InputConnectedField.tsx"] = {
    name: "InputConnectedField",
    docgenInfo: InputConnectedField.__docgenInfo,
    path: "src/components/form/components/InputConnectedField.tsx"
  };
}

export { InputConnectedField };
