// gatsby-interface/dist/src/components/form/components/SelectConnectedField.esm.js
import { slicedToArray as _slicedToArray, extends as _extends } from '../../../../_virtual/_rollupPluginBabelHelpers.js';
import { jsx } from '@emotion/core';
import React__default from 'react';
import _pt from 'prop-types';
import { useConnectedField } from '../hooks/useConnectedField.esm.js';
import { SelectFieldBlock } from './SelectFieldBlock.esm.js';

var SelectConnectedField = React__default.forwardRef(function SelectConnectedField(props, ref) {
  var _useConnectedField = useConnectedField(props.name),
      _useConnectedField2 = _slicedToArray(_useConnectedField, 1),
      connectedProps = _useConnectedField2[0];

  return jsx(SelectFieldBlock, _extends({
    ref: ref
  }, connectedProps, props));
});
SelectConnectedField.propTypes = {
  name: _pt.string.isRequired,
  id: _pt.string,
  label: _pt.node
};
SelectConnectedField.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "SelectConnectedField"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form/components/SelectConnectedField.tsx"] = {
    name: "SelectConnectedField",
    docgenInfo: SelectConnectedField.__docgenInfo,
    path: "src/components/form/components/SelectConnectedField.tsx"
  };
}

export { SelectConnectedField };
