// gatsby-interface/dist/src/components/form/components/SelectField.esm.js
import { extends as _extends, objectWithoutProperties as _objectWithoutProperties } from '../../../../_virtual/_rollupPluginBabelHelpers.js';
import { jsx } from '@emotion/core';
import React__default from 'react';
import { showFormFieldDeprecatedMessage } from '../../../utils/maintenance/deprecationMessages.esm.js';
import { getInputStyles } from './FormField.helpers.esm.js';
import { useStyledFieldLabel, useStyledFieldError, useStyledFieldHint } from './FormField.esm.js';
import { useFormFieldSkeleton } from '../../form-skeletons/components/FormFieldSkeleton.esm.js';
import { SelectFieldSkeleton, SelectFieldSkeletonControl, SelectFieldSkeletonLabel, SelectFieldSkeletonError, SelectFieldSkeletonHint } from '../../form-skeletons/components/SelectFieldSkeleton.esm.js';

function SelectField(props) {
  showFormFieldDeprecatedMessage("SelectField");
  return jsx(SelectFieldSkeleton, props);
}
var SelectFieldControl = React__default.forwardRef(function SelectFieldControl(props, ref) {
  var _useFormFieldSkeleton = useFormFieldSkeleton(),
      hasError = _useFormFieldSkeleton.hasError;

  return jsx(SelectFieldSkeletonControl, _extends({
    ref: ref,
    css: function css(theme) {
      return [getInputStyles(theme, hasError), {
        padding: "0 ".concat(theme.space[3]),
        backgroundImage: "url(\"data:image/svg+xml,%3Csvg stroke='currentColor' fill='currentColor' stroke-width='0' viewBox='0 0 24 24' height='1em' width='1em' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath d='M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z'%3E%3C/path%3E%3C/svg%3E\")",
        backgroundRepeat: "no-repeat",
        backgroundPosition: "right ".concat(theme.space[3], " top 50%, 0 0")
      }];
    }
  }, props));
});
function SelectFieldLabel(_ref) {
  var children = _ref.children,
      size = _ref.size,
      isRequired = _ref.isRequired,
      props = _objectWithoutProperties(_ref, ["children", "size", "isRequired"]);

  var styledProps = useStyledFieldLabel(children, {
    size: size,
    isRequired: isRequired
  });
  return jsx(SelectFieldSkeletonLabel, _extends({}, props, styledProps));
}
function SelectFieldError(_ref2) {
  var children = _ref2.children,
      props = _objectWithoutProperties(_ref2, ["children"]);

  var styledProps = useStyledFieldError(children);
  return jsx(SelectFieldSkeletonError, _extends({}, props, styledProps));
}
function SelectFieldHint(props) {
  var styledProps = useStyledFieldHint();
  return jsx(SelectFieldSkeletonHint, _extends({}, props, styledProps));
}
SelectField.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "SelectField"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form/components/SelectField.tsx"] = {
    name: "SelectField",
    docgenInfo: SelectField.__docgenInfo,
    path: "src/components/form/components/SelectField.tsx"
  };
}

SelectFieldControl.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "SelectFieldControl"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form/components/SelectField.tsx"] = {
    name: "SelectFieldControl",
    docgenInfo: SelectFieldControl.__docgenInfo,
    path: "src/components/form/components/SelectField.tsx"
  };
}

SelectFieldLabel.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "SelectFieldLabel"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form/components/SelectField.tsx"] = {
    name: "SelectFieldLabel",
    docgenInfo: SelectFieldLabel.__docgenInfo,
    path: "src/components/form/components/SelectField.tsx"
  };
}

SelectFieldError.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "SelectFieldError"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form/components/SelectField.tsx"] = {
    name: "SelectFieldError",
    docgenInfo: SelectFieldError.__docgenInfo,
    path: "src/components/form/components/SelectField.tsx"
  };
}

SelectFieldHint.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "SelectFieldHint"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form/components/SelectField.tsx"] = {
    name: "SelectFieldHint",
    docgenInfo: SelectFieldHint.__docgenInfo,
    path: "src/components/form/components/SelectField.tsx"
  };
}

export { SelectField, SelectFieldControl, SelectFieldError, SelectFieldHint, SelectFieldLabel };
