// gatsby-interface/dist/src/components/form/components/SelectFieldBlock.esm.js
import { objectWithoutProperties as _objectWithoutProperties, extends as _extends } from '../../../../_virtual/_rollupPluginBabelHelpers.js';
import { jsx } from '@emotion/core';
import React__default from 'react';
import { FormFieldBlock } from './FormFieldBlock.esm.js';
import { StyledSelect } from './styled-primitives/StyledSelect.esm.js';

var SelectFieldBlock = React__default.forwardRef(function SelectFieldBlock(props, ref) {
  var id = props.id,
      label = props.label,
      labelSize = props.labelSize,
      error = props.error,
      hint = props.hint,
      className = props.className,
      validationMode = props.validationMode,
      layout = props.layout,
      required = props.required,
      rest = _objectWithoutProperties(props, ["id", "label", "labelSize", "error", "hint", "className", "validationMode", "layout", "required"]);

  return jsx(FormFieldBlock, {
    id: id,
    label: label,
    error: error,
    hint: hint,
    required: required,
    labelSize: labelSize,
    validationMode: validationMode,
    layout: layout,
    className: className
  }, function (controlProps) {
    return jsx(StyledSelect, _extends({
      ref: ref
    }, controlProps, rest));
  });
});
SelectFieldBlock.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "SelectFieldBlock"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form/components/SelectFieldBlock.tsx"] = {
    name: "SelectFieldBlock",
    docgenInfo: SelectFieldBlock.__docgenInfo,
    path: "src/components/form/components/SelectFieldBlock.tsx"
  };
}

export { SelectFieldBlock };
