// gatsby-interface/dist/src/components/form/components/TextAreaFieldBlock.esm.js
import { objectWithoutProperties as _objectWithoutProperties, extends as _extends } from '../../../../_virtual/_rollupPluginBabelHelpers.js';
import { jsx } from '@emotion/core';
import React__default from 'react';
import { FormFieldBlock } from './FormFieldBlock.esm.js';
import { StyledTextArea } from './styled-primitives/StyledTextArea.esm.js';

var TextAreaFieldBlock = React__default.forwardRef(function TextAreaFieldBlock(props, ref) {
  var id = props.id,
      label = props.label,
      labelSize = props.labelSize,
      error = props.error,
      hint = props.hint,
      className = props.className,
      validationMode = props.validationMode,
      layout = props.layout,
      required = props.required,
      rest = _objectWithoutProperties(props, ["id", "label", "labelSize", "error", "hint", "className", "validationMode", "layout", "required"]);

  return jsx(FormFieldBlock, {
    id: id,
    label: label,
    error: error,
    hint: hint,
    required: required,
    labelSize: labelSize,
    validationMode: validationMode,
    layout: layout,
    className: className
  }, function (controlProps) {
    return jsx(StyledTextArea, _extends({
      ref: ref
    }, controlProps, rest));
  });
});
TextAreaFieldBlock.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "TextAreaFieldBlock"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form/components/TextAreaFieldBlock.tsx"] = {
    name: "TextAreaFieldBlock",
    docgenInfo: TextAreaFieldBlock.__docgenInfo,
    path: "src/components/form/components/TextAreaFieldBlock.tsx"
  };
}

export { TextAreaFieldBlock };
