// gatsby-interface/dist/src/components/form/components/TextAreaConnectedField.esm.js
import { slicedToArray as _slicedToArray, extends as _extends } from '../../../../_virtual/_rollupPluginBabelHelpers.js';
import { jsx } from '@emotion/core';
import React__default from 'react';
import _pt from 'prop-types';
import { useConnectedField } from '../hooks/useConnectedField.esm.js';
import { TextAreaFieldBlock } from './TextAreaFieldBlock.esm.js';

var TextAreaConnectedField = React__default.forwardRef(function TextAreaConnectedField(props, ref) {
  var _useConnectedField = useConnectedField(props.name),
      _useConnectedField2 = _slicedToArray(_useConnectedField, 1),
      connectedProps = _useConnectedField2[0];

  return jsx(TextAreaFieldBlock, _extends({
    ref: ref
  }, connectedProps, props));
});
TextAreaConnectedField.propTypes = {
  name: _pt.string.isRequired,
  id: _pt.string,
  label: _pt.node
};
TextAreaConnectedField.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "TextAreaConnectedField"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form/components/TextAreaConnectedField.tsx"] = {
    name: "TextAreaConnectedField",
    docgenInfo: TextAreaConnectedField.__docgenInfo,
    path: "src/components/form/components/TextAreaConnectedField.tsx"
  };
}

export { TextAreaConnectedField };
