// gatsby-core-utils/dist/index.js
"use strict";

exports.__esModule = true;
var _exportNames = {
  createContentDigest: true,
  isNodeInternalModulePath: true,
  joinPath: true,
  slash: true,
  cpuCoreCount: true,
  urlResolve: true,
  getCIName: true,
  isCI: true,
  createRequireFromPath: true,
  getConfigStore: true,
  getGatsbyVersion: true
};
exports.getGatsbyVersion = exports.getConfigStore = exports.createRequireFromPath = exports.isCI = exports.getCIName = exports.urlResolve = exports.cpuCoreCount = exports.slash = exports.joinPath = exports.isNodeInternalModulePath = exports.createContentDigest = void 0;

var _createContentDigest = require("./create-content-digest");

exports.createContentDigest = _createContentDigest.createContentDigest;

var _path = require("./path");

exports.isNodeInternalModulePath = _path.isNodeInternalModulePath;
exports.joinPath = _path.joinPath;
exports.slash = _path.slash;

var _cpuCoreCount = require("./cpu-core-count");

exports.cpuCoreCount = _cpuCoreCount.cpuCoreCount;

var _url = require("./url");

exports.urlResolve = _url.urlResolve;

var _ci = require("./ci");

exports.getCIName = _ci.getCIName;
exports.isCI = _ci.isCI;

var _createRequireFromPath = require("./create-require-from-path");

exports.createRequireFromPath = _createRequireFromPath.createRequireFromPath;

var _getConfigStore = require("./get-config-store");

exports.getConfigStore = _getConfigStore.getConfigStore;

var _getGatsbyVersion = require("./get-gatsby-version");

exports.getGatsbyVersion = _getGatsbyVersion.getGatsbyVersion;

var _serviceLock = require("./service-lock");

Object.keys(_serviceLock).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  exports[key] = _serviceLock[key];
});

var _siteMetadata = require("./site-metadata");

Object.keys(_siteMetadata).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  exports[key] = _siteMetadata[key];
});
