// gatsby-interface/dist/src/components/form-skeletons/components/CheckboxFieldSkeleton.esm.js
import { extends as _extends } from '../../../../_virtual/_rollupPluginBabelHelpers.js';
import React__default from 'react';
import { showFormSkeletonDeprecatedMessage } from '../../../utils/maintenance/deprecationMessages.esm.js';
import { getFinalAriaDescribedBy } from '../utils/index.esm.js';
import { FormFieldSkeleton, useFormFieldSkeleton, FormFieldSkeletonLabel, FormFieldSkeletonHint, FormFieldSkeletonError } from './FormFieldSkeleton.esm.js';

function CheckboxFieldSkeleton(props) {
  showFormSkeletonDeprecatedMessage("CheckboxFieldSkeleton");
  return /*#__PURE__*/React__default.createElement(FormFieldSkeleton, props);
}
var CheckboxFieldSkeletonControl = React__default.forwardRef(function CheckboxFieldSkeletonControl(props, ref) {
  var _useFormFieldSkeleton = useFormFieldSkeleton(),
      id = _useFormFieldSkeleton.id,
      hasError = _useFormFieldSkeleton.hasError,
      meta = _useFormFieldSkeleton.meta;

  return /*#__PURE__*/React__default.createElement("input", _extends({
    id: id
  }, props, {
    type: "checkbox",
    "aria-describedby": getFinalAriaDescribedBy(meta.controlDescribedBy, props["aria-describedby"]),
    "aria-invalid": hasError,
    ref: ref
  }));
});
function CheckboxFieldSkeletonLabel(props) {
  return /*#__PURE__*/React__default.createElement(FormFieldSkeletonLabel, props);
}
function CheckboxFieldSkeletonHint(props) {
  return /*#__PURE__*/React__default.createElement(FormFieldSkeletonHint, props);
}
function CheckboxFieldSkeletonError(props) {
  return /*#__PURE__*/React__default.createElement(FormFieldSkeletonError, props);
}
CheckboxFieldSkeleton.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "CheckboxFieldSkeleton"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form-skeletons/components/CheckboxFieldSkeleton.tsx"] = {
    name: "CheckboxFieldSkeleton",
    docgenInfo: CheckboxFieldSkeleton.__docgenInfo,
    path: "src/components/form-skeletons/components/CheckboxFieldSkeleton.tsx"
  };
}

CheckboxFieldSkeletonControl.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "CheckboxFieldSkeletonControl"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form-skeletons/components/CheckboxFieldSkeleton.tsx"] = {
    name: "CheckboxFieldSkeletonControl",
    docgenInfo: CheckboxFieldSkeletonControl.__docgenInfo,
    path: "src/components/form-skeletons/components/CheckboxFieldSkeleton.tsx"
  };
}

CheckboxFieldSkeletonLabel.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "CheckboxFieldSkeletonLabel"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form-skeletons/components/CheckboxFieldSkeleton.tsx"] = {
    name: "CheckboxFieldSkeletonLabel",
    docgenInfo: CheckboxFieldSkeletonLabel.__docgenInfo,
    path: "src/components/form-skeletons/components/CheckboxFieldSkeleton.tsx"
  };
}

CheckboxFieldSkeletonHint.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "CheckboxFieldSkeletonHint"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form-skeletons/components/CheckboxFieldSkeleton.tsx"] = {
    name: "CheckboxFieldSkeletonHint",
    docgenInfo: CheckboxFieldSkeletonHint.__docgenInfo,
    path: "src/components/form-skeletons/components/CheckboxFieldSkeleton.tsx"
  };
}

CheckboxFieldSkeletonError.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "CheckboxFieldSkeletonError"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form-skeletons/components/CheckboxFieldSkeleton.tsx"] = {
    name: "CheckboxFieldSkeletonError",
    docgenInfo: CheckboxFieldSkeletonError.__docgenInfo,
    path: "src/components/form-skeletons/components/CheckboxFieldSkeleton.tsx"
  };
}

export { CheckboxFieldSkeleton, CheckboxFieldSkeletonControl, CheckboxFieldSkeletonError, CheckboxFieldSkeletonHint, CheckboxFieldSkeletonLabel };
