// gatsby-interface/dist/src/components/form-skeletons/components/CheckboxGroupFieldSkeleton.esm.js
import { extends as _extends } from '../../../../_virtual/_rollupPluginBabelHelpers.js';
import React__default from 'react';
import { showFormSkeletonDeprecatedMessage } from '../../../utils/maintenance/deprecationMessages.esm.js';
import { FormFieldSkeletonHint, FormFieldSkeletonError } from './FormFieldSkeleton.esm.js';
import { FormGroupFieldSkeleton, FormGroupFieldSkeletonLabel, FormGroupFieldSkeletonOption, FormGroupFieldSkeletonOptionLabel } from './FormGroupFieldSkeleton.esm.js';

function CheckboxGroupFieldSkeleton(props) {
  showFormSkeletonDeprecatedMessage("CheckboxGroupFieldSkeleton", true);
  return /*#__PURE__*/React__default.createElement(FormGroupFieldSkeleton, props);
}
function CheckboxGroupFieldSkeletonLabel(props) {
  return /*#__PURE__*/React__default.createElement(FormGroupFieldSkeletonLabel, props);
}
function CheckboxGroupFieldSkeletonHint(props) {
  return /*#__PURE__*/React__default.createElement(FormFieldSkeletonHint, props);
}
function CheckboxGroupFieldSkeletonError(props) {
  return /*#__PURE__*/React__default.createElement(FormFieldSkeletonError, props);
}
var CheckboxGroupFieldSkeletonOption = React__default.forwardRef(function CheckboxGroupFieldSkeletonOption(props, ref) {
  return /*#__PURE__*/React__default.createElement(FormGroupFieldSkeletonOption, _extends({
    ref: ref,
    type: "checkbox"
  }, props));
});
function CheckboxGroupFieldSkeletonOptionLabel(props) {
  return /*#__PURE__*/React__default.createElement(FormGroupFieldSkeletonOptionLabel, props);
}
CheckboxGroupFieldSkeleton.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "CheckboxGroupFieldSkeleton"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form-skeletons/components/CheckboxGroupFieldSkeleton.tsx"] = {
    name: "CheckboxGroupFieldSkeleton",
    docgenInfo: CheckboxGroupFieldSkeleton.__docgenInfo,
    path: "src/components/form-skeletons/components/CheckboxGroupFieldSkeleton.tsx"
  };
}

CheckboxGroupFieldSkeletonLabel.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "CheckboxGroupFieldSkeletonLabel"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form-skeletons/components/CheckboxGroupFieldSkeleton.tsx"] = {
    name: "CheckboxGroupFieldSkeletonLabel",
    docgenInfo: CheckboxGroupFieldSkeletonLabel.__docgenInfo,
    path: "src/components/form-skeletons/components/CheckboxGroupFieldSkeleton.tsx"
  };
}

CheckboxGroupFieldSkeletonHint.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "CheckboxGroupFieldSkeletonHint"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form-skeletons/components/CheckboxGroupFieldSkeleton.tsx"] = {
    name: "CheckboxGroupFieldSkeletonHint",
    docgenInfo: CheckboxGroupFieldSkeletonHint.__docgenInfo,
    path: "src/components/form-skeletons/components/CheckboxGroupFieldSkeleton.tsx"
  };
}

CheckboxGroupFieldSkeletonError.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "CheckboxGroupFieldSkeletonError"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form-skeletons/components/CheckboxGroupFieldSkeleton.tsx"] = {
    name: "CheckboxGroupFieldSkeletonError",
    docgenInfo: CheckboxGroupFieldSkeletonError.__docgenInfo,
    path: "src/components/form-skeletons/components/CheckboxGroupFieldSkeleton.tsx"
  };
}

CheckboxGroupFieldSkeletonOption.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "CheckboxGroupFieldSkeletonOption"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form-skeletons/components/CheckboxGroupFieldSkeleton.tsx"] = {
    name: "CheckboxGroupFieldSkeletonOption",
    docgenInfo: CheckboxGroupFieldSkeletonOption.__docgenInfo,
    path: "src/components/form-skeletons/components/CheckboxGroupFieldSkeleton.tsx"
  };
}

CheckboxGroupFieldSkeletonOptionLabel.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "CheckboxGroupFieldSkeletonOptionLabel"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form-skeletons/components/CheckboxGroupFieldSkeleton.tsx"] = {
    name: "CheckboxGroupFieldSkeletonOptionLabel",
    docgenInfo: CheckboxGroupFieldSkeletonOptionLabel.__docgenInfo,
    path: "src/components/form-skeletons/components/CheckboxGroupFieldSkeleton.tsx"
  };
}

export { CheckboxGroupFieldSkeleton, CheckboxGroupFieldSkeletonError, CheckboxGroupFieldSkeletonHint, CheckboxGroupFieldSkeletonLabel, CheckboxGroupFieldSkeletonOption, CheckboxGroupFieldSkeletonOptionLabel };
