// gatsby-interface/dist/src/components/form-skeletons/components/InputFieldSkeleton.esm.js
import { extends as _extends } from '../../../../_virtual/_rollupPluginBabelHelpers.js';
import React__default from 'react';
import { showFormSkeletonDeprecatedMessage } from '../../../utils/maintenance/deprecationMessages.esm.js';
import { getFinalAriaDescribedBy } from '../utils/index.esm.js';
import { FormFieldSkeleton, useFormFieldSkeleton, FormFieldSkeletonLabel, FormFieldSkeletonHint, FormFieldSkeletonError } from './FormFieldSkeleton.esm.js';

function InputFieldSkeleton(props) {
  showFormSkeletonDeprecatedMessage("InputFieldSkeleton");
  return /*#__PURE__*/React__default.createElement(FormFieldSkeleton, props);
}
var InputFieldSkeletonControl = React__default.forwardRef(function InputFieldSkeletonControl(props, ref) {
  var _useFormFieldSkeleton = useFormFieldSkeleton(),
      id = _useFormFieldSkeleton.id,
      hasError = _useFormFieldSkeleton.hasError,
      meta = _useFormFieldSkeleton.meta;

  return /*#__PURE__*/React__default.createElement("input", _extends({
    id: id
  }, props, {
    "aria-describedby": getFinalAriaDescribedBy(meta.controlDescribedBy, props["aria-describedby"]),
    "aria-invalid": hasError,
    ref: ref
  }));
});
function InputFieldSkeletonLabel(props) {
  return /*#__PURE__*/React__default.createElement(FormFieldSkeletonLabel, props);
}
function InputFieldSkeletonHint(props) {
  return /*#__PURE__*/React__default.createElement(FormFieldSkeletonHint, props);
}
function InputFieldSkeletonError(props) {
  return /*#__PURE__*/React__default.createElement(FormFieldSkeletonError, props);
}
InputFieldSkeleton.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "InputFieldSkeleton"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form-skeletons/components/InputFieldSkeleton.tsx"] = {
    name: "InputFieldSkeleton",
    docgenInfo: InputFieldSkeleton.__docgenInfo,
    path: "src/components/form-skeletons/components/InputFieldSkeleton.tsx"
  };
}

InputFieldSkeletonControl.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "InputFieldSkeletonControl"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form-skeletons/components/InputFieldSkeleton.tsx"] = {
    name: "InputFieldSkeletonControl",
    docgenInfo: InputFieldSkeletonControl.__docgenInfo,
    path: "src/components/form-skeletons/components/InputFieldSkeleton.tsx"
  };
}

InputFieldSkeletonLabel.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "InputFieldSkeletonLabel"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form-skeletons/components/InputFieldSkeleton.tsx"] = {
    name: "InputFieldSkeletonLabel",
    docgenInfo: InputFieldSkeletonLabel.__docgenInfo,
    path: "src/components/form-skeletons/components/InputFieldSkeleton.tsx"
  };
}

InputFieldSkeletonHint.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "InputFieldSkeletonHint"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form-skeletons/components/InputFieldSkeleton.tsx"] = {
    name: "InputFieldSkeletonHint",
    docgenInfo: InputFieldSkeletonHint.__docgenInfo,
    path: "src/components/form-skeletons/components/InputFieldSkeleton.tsx"
  };
}

InputFieldSkeletonError.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "InputFieldSkeletonError"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form-skeletons/components/InputFieldSkeleton.tsx"] = {
    name: "InputFieldSkeletonError",
    docgenInfo: InputFieldSkeletonError.__docgenInfo,
    path: "src/components/form-skeletons/components/InputFieldSkeleton.tsx"
  };
}

export { InputFieldSkeleton, InputFieldSkeletonControl, InputFieldSkeletonError, InputFieldSkeletonHint, InputFieldSkeletonLabel };
