// gatsby-interface/dist/src/components/form-skeletons/components/RadioButtonFieldSkeleton.esm.js
import { extends as _extends } from '../../../../_virtual/_rollupPluginBabelHelpers.js';
import React__default from 'react';
import { showFormSkeletonDeprecatedMessage } from '../../../utils/maintenance/deprecationMessages.esm.js';
import { FormFieldSkeletonHint, FormFieldSkeletonError } from './FormFieldSkeleton.esm.js';
import { FormGroupFieldSkeleton, FormGroupFieldSkeletonLabel, FormGroupFieldSkeletonOption, FormGroupFieldSkeletonOptionLabel } from './FormGroupFieldSkeleton.esm.js';

function RadioButtonFieldSkeleton(props) {
  showFormSkeletonDeprecatedMessage("RadioButtonFieldSkeleton", true);
  return /*#__PURE__*/React__default.createElement(FormGroupFieldSkeleton, props);
}
function RadioButtonFieldSkeletonLabel(props) {
  return /*#__PURE__*/React__default.createElement(FormGroupFieldSkeletonLabel, props);
}
function RadioButtonFieldSkeletonHint(props) {
  return /*#__PURE__*/React__default.createElement(FormFieldSkeletonHint, props);
}
function RadioButtonFieldSkeletonError(props) {
  return /*#__PURE__*/React__default.createElement(FormFieldSkeletonError, props);
}
var RadioButtonFieldSkeletonOption = React__default.forwardRef(function RadioButtonFieldSkeletonOption(props, ref) {
  return /*#__PURE__*/React__default.createElement(FormGroupFieldSkeletonOption, _extends({
    ref: ref,
    type: "radio"
  }, props));
});
function RadioButtonFieldSkeletonOptionLabel(props) {
  return /*#__PURE__*/React__default.createElement(FormGroupFieldSkeletonOptionLabel, props);
}
RadioButtonFieldSkeleton.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "RadioButtonFieldSkeleton"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form-skeletons/components/RadioButtonFieldSkeleton.tsx"] = {
    name: "RadioButtonFieldSkeleton",
    docgenInfo: RadioButtonFieldSkeleton.__docgenInfo,
    path: "src/components/form-skeletons/components/RadioButtonFieldSkeleton.tsx"
  };
}

RadioButtonFieldSkeletonLabel.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "RadioButtonFieldSkeletonLabel"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form-skeletons/components/RadioButtonFieldSkeleton.tsx"] = {
    name: "RadioButtonFieldSkeletonLabel",
    docgenInfo: RadioButtonFieldSkeletonLabel.__docgenInfo,
    path: "src/components/form-skeletons/components/RadioButtonFieldSkeleton.tsx"
  };
}

RadioButtonFieldSkeletonHint.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "RadioButtonFieldSkeletonHint"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form-skeletons/components/RadioButtonFieldSkeleton.tsx"] = {
    name: "RadioButtonFieldSkeletonHint",
    docgenInfo: RadioButtonFieldSkeletonHint.__docgenInfo,
    path: "src/components/form-skeletons/components/RadioButtonFieldSkeleton.tsx"
  };
}

RadioButtonFieldSkeletonError.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "RadioButtonFieldSkeletonError"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form-skeletons/components/RadioButtonFieldSkeleton.tsx"] = {
    name: "RadioButtonFieldSkeletonError",
    docgenInfo: RadioButtonFieldSkeletonError.__docgenInfo,
    path: "src/components/form-skeletons/components/RadioButtonFieldSkeleton.tsx"
  };
}

RadioButtonFieldSkeletonOption.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "RadioButtonFieldSkeletonOption"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form-skeletons/components/RadioButtonFieldSkeleton.tsx"] = {
    name: "RadioButtonFieldSkeletonOption",
    docgenInfo: RadioButtonFieldSkeletonOption.__docgenInfo,
    path: "src/components/form-skeletons/components/RadioButtonFieldSkeleton.tsx"
  };
}

RadioButtonFieldSkeletonOptionLabel.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "RadioButtonFieldSkeletonOptionLabel"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form-skeletons/components/RadioButtonFieldSkeleton.tsx"] = {
    name: "RadioButtonFieldSkeletonOptionLabel",
    docgenInfo: RadioButtonFieldSkeletonOptionLabel.__docgenInfo,
    path: "src/components/form-skeletons/components/RadioButtonFieldSkeleton.tsx"
  };
}

export { RadioButtonFieldSkeleton, RadioButtonFieldSkeletonError, RadioButtonFieldSkeletonHint, RadioButtonFieldSkeletonLabel, RadioButtonFieldSkeletonOption, RadioButtonFieldSkeletonOptionLabel };
