// gatsby-interface/dist/src/components/form-skeletons/components/SelectFieldSkeleton.esm.js
import { objectWithoutProperties as _objectWithoutProperties, extends as _extends } from '../../../../_virtual/_rollupPluginBabelHelpers.js';
import React__default from 'react';
import _pt from 'prop-types';
import { showFormSkeletonDeprecatedMessage } from '../../../utils/maintenance/deprecationMessages.esm.js';
import { getFinalAriaDescribedBy } from '../utils/index.esm.js';
import { FormFieldSkeleton, useFormFieldSkeleton, FormFieldSkeletonLabel, FormFieldSkeletonHint, FormFieldSkeletonError } from './FormFieldSkeleton.esm.js';

function SelectFieldSkeleton(props) {
  showFormSkeletonDeprecatedMessage("SelectFieldSkeleton");
  return /*#__PURE__*/React__default.createElement(FormFieldSkeleton, props);
}
var SelectFieldSkeletonControl = React__default.forwardRef(function SelectFieldSkeletonControl(_ref, ref) {
  var options = _ref.options,
      rest = _objectWithoutProperties(_ref, ["options"]);

  var _useFormFieldSkeleton = useFormFieldSkeleton(),
      id = _useFormFieldSkeleton.id,
      hasError = _useFormFieldSkeleton.hasError,
      meta = _useFormFieldSkeleton.meta;

  return /*#__PURE__*/React__default.createElement("select", _extends({
    id: id
  }, rest, {
    "aria-describedby": getFinalAriaDescribedBy(meta.controlDescribedBy, rest["aria-describedby"]),
    "aria-invalid": hasError,
    ref: ref
  }), options.map(renderOption));
});
SelectFieldSkeletonControl.propTypes = {
  options: _pt.arrayOf(_pt.shape({
    value: _pt.string.isRequired,
    label: _pt.string.isRequired
  })).isRequired
};

function renderOption(_ref2) {
  var label = _ref2.label,
      value = _ref2.value;
  return /*#__PURE__*/React__default.createElement("option", {
    key: value,
    value: value
  }, label);
}

function SelectFieldSkeletonLabel(props) {
  return /*#__PURE__*/React__default.createElement(FormFieldSkeletonLabel, props);
}
function SelectFieldSkeletonHint(props) {
  return /*#__PURE__*/React__default.createElement(FormFieldSkeletonHint, props);
}
function SelectFieldSkeletonError(props) {
  return /*#__PURE__*/React__default.createElement(FormFieldSkeletonError, props);
}
SelectFieldSkeleton.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "SelectFieldSkeleton"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form-skeletons/components/SelectFieldSkeleton.tsx"] = {
    name: "SelectFieldSkeleton",
    docgenInfo: SelectFieldSkeleton.__docgenInfo,
    path: "src/components/form-skeletons/components/SelectFieldSkeleton.tsx"
  };
}

SelectFieldSkeletonControl.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "SelectFieldSkeletonControl"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form-skeletons/components/SelectFieldSkeleton.tsx"] = {
    name: "SelectFieldSkeletonControl",
    docgenInfo: SelectFieldSkeletonControl.__docgenInfo,
    path: "src/components/form-skeletons/components/SelectFieldSkeleton.tsx"
  };
}

SelectFieldSkeletonLabel.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "SelectFieldSkeletonLabel"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form-skeletons/components/SelectFieldSkeleton.tsx"] = {
    name: "SelectFieldSkeletonLabel",
    docgenInfo: SelectFieldSkeletonLabel.__docgenInfo,
    path: "src/components/form-skeletons/components/SelectFieldSkeleton.tsx"
  };
}

SelectFieldSkeletonHint.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "SelectFieldSkeletonHint"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form-skeletons/components/SelectFieldSkeleton.tsx"] = {
    name: "SelectFieldSkeletonHint",
    docgenInfo: SelectFieldSkeletonHint.__docgenInfo,
    path: "src/components/form-skeletons/components/SelectFieldSkeleton.tsx"
  };
}

SelectFieldSkeletonError.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "SelectFieldSkeletonError"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form-skeletons/components/SelectFieldSkeleton.tsx"] = {
    name: "SelectFieldSkeletonError",
    docgenInfo: SelectFieldSkeletonError.__docgenInfo,
    path: "src/components/form-skeletons/components/SelectFieldSkeleton.tsx"
  };
}

export { SelectFieldSkeleton, SelectFieldSkeletonControl, SelectFieldSkeletonError, SelectFieldSkeletonHint, SelectFieldSkeletonLabel };
