// gatsby-interface/dist/src/components/form-skeletons/components/TextAreaFieldSkeleton.esm.js
import { extends as _extends } from '../../../../_virtual/_rollupPluginBabelHelpers.js';
import React__default from 'react';
import { showFormSkeletonDeprecatedMessage } from '../../../utils/maintenance/deprecationMessages.esm.js';
import { getFinalAriaDescribedBy } from '../utils/index.esm.js';
import { FormFieldSkeleton, useFormFieldSkeleton, FormFieldSkeletonLabel, FormFieldSkeletonHint, FormFieldSkeletonError } from './FormFieldSkeleton.esm.js';

function TextAreaFieldSkeleton(props) {
  showFormSkeletonDeprecatedMessage("TextAreaFieldSkeleton");
  return /*#__PURE__*/React__default.createElement(FormFieldSkeleton, props);
}
var TextAreaFieldSkeletonControl = React__default.forwardRef(function TextAreaFieldSkeletonControl(props, ref) {
  var _useFormFieldSkeleton = useFormFieldSkeleton(),
      id = _useFormFieldSkeleton.id,
      hasError = _useFormFieldSkeleton.hasError,
      meta = _useFormFieldSkeleton.meta;

  return /*#__PURE__*/React__default.createElement("textarea", _extends({
    id: id
  }, props, {
    "aria-describedby": getFinalAriaDescribedBy(meta.controlDescribedBy, props["aria-describedby"]),
    "aria-invalid": hasError,
    ref: ref
  }));
});
function TextAreaFieldSkeletonLabel(props) {
  return /*#__PURE__*/React__default.createElement(FormFieldSkeletonLabel, props);
}
function TextAreaFieldSkeletonHint(props) {
  return /*#__PURE__*/React__default.createElement(FormFieldSkeletonHint, props);
}
function TextAreaFieldSkeletonError(props) {
  return /*#__PURE__*/React__default.createElement(FormFieldSkeletonError, props);
}
TextAreaFieldSkeleton.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "TextAreaFieldSkeleton"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form-skeletons/components/TextAreaFieldSkeleton.tsx"] = {
    name: "TextAreaFieldSkeleton",
    docgenInfo: TextAreaFieldSkeleton.__docgenInfo,
    path: "src/components/form-skeletons/components/TextAreaFieldSkeleton.tsx"
  };
}

TextAreaFieldSkeletonControl.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "TextAreaFieldSkeletonControl"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form-skeletons/components/TextAreaFieldSkeleton.tsx"] = {
    name: "TextAreaFieldSkeletonControl",
    docgenInfo: TextAreaFieldSkeletonControl.__docgenInfo,
    path: "src/components/form-skeletons/components/TextAreaFieldSkeleton.tsx"
  };
}

TextAreaFieldSkeletonLabel.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "TextAreaFieldSkeletonLabel"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form-skeletons/components/TextAreaFieldSkeleton.tsx"] = {
    name: "TextAreaFieldSkeletonLabel",
    docgenInfo: TextAreaFieldSkeletonLabel.__docgenInfo,
    path: "src/components/form-skeletons/components/TextAreaFieldSkeleton.tsx"
  };
}

TextAreaFieldSkeletonHint.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "TextAreaFieldSkeletonHint"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form-skeletons/components/TextAreaFieldSkeleton.tsx"] = {
    name: "TextAreaFieldSkeletonHint",
    docgenInfo: TextAreaFieldSkeletonHint.__docgenInfo,
    path: "src/components/form-skeletons/components/TextAreaFieldSkeleton.tsx"
  };
}

TextAreaFieldSkeletonError.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "TextAreaFieldSkeletonError"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/form-skeletons/components/TextAreaFieldSkeleton.tsx"] = {
    name: "TextAreaFieldSkeletonError",
    docgenInfo: TextAreaFieldSkeletonError.__docgenInfo,
    path: "src/components/form-skeletons/components/TextAreaFieldSkeleton.tsx"
  };
}

export { TextAreaFieldSkeleton, TextAreaFieldSkeletonControl, TextAreaFieldSkeletonError, TextAreaFieldSkeletonHint, TextAreaFieldSkeletonLabel };
