// node-object-hash/dist/typeGuess.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Type mapping rules.
 */
exports.TYPE_MAP = {
    Array: 'array',
    Int8Array: 'array',
    Uint8Array: 'array',
    Uint8ClampedArray: 'array',
    Int16Array: 'array',
    Uint16Array: 'array',
    Int32Array: 'array',
    Uint32Array: 'array',
    Float32Array: 'array',
    Float64Array: 'array',
    Buffer: 'array',
    Map: 'map',
    Set: 'set',
    Date: 'date',
    String: 'string',
    Number: 'number',
    Boolean: 'boolean',
    Object: 'object',
};
/**
 * Guess object type
 * @param obj analyzed object
 * @returns object type
 */
function guessObjectType(obj) {
    if (obj === null) {
        return 'null';
    }
    var type = obj.constructor ? obj.constructor.name : 'unknown';
    return exports.TYPE_MAP[type] || 'unknown';
}
exports.guessObjectType = guessObjectType;
/**
 * Guess variable type
 * @param obj analyzed variable
 * @returns variable type
 */
function guessType(obj) {
    var type = typeof obj;
    return type !== 'object' ? type : guessObjectType(obj);
}
exports.guessType = guessType;
