// gatsby-interface/dist/src/components/NumberBadge/NumberBadge.esm.js
import { objectWithoutProperties as _objectWithoutProperties, extends as _extends } from '../../../_virtual/_rollupPluginBabelHelpers.js';
import { jsx } from '@emotion/core';
import _pt from 'prop-types';

var baseCss = function baseCss(theme) {
  return {
    display: "inline-block",
    fontFamily: theme.fonts.sans,
    fontSize: theme.fontSizes[0],
    lineHeight: theme.lineHeights.solid,
    fontWeight: theme.fontWeights.semiBold,
    backgroundColor: theme.colors.grey[20],
    color: theme.colors.grey[60],
    paddingLeft: "calc(3 * ".concat(theme.space[1], ")"),
    paddingRight: "calc(3 * ".concat(theme.space[1], ")"),
    paddingTop: theme.space[1],
    paddingBottom: theme.space[1],
    borderRadius: "10px"
  };
};

var toneCss = {
  DANGER: function DANGER(theme) {
    return {
      backgroundColor: theme.tones.DANGER.lighter,
      color: theme.tones.DANGER.superDark
    };
  },
  WARNING: function WARNING(theme) {
    return {
      backgroundColor: theme.tones.WARNING.light,
      color: theme.tones.WARNING.textInverted
    };
  },
  NEUTRAL: function NEUTRAL(theme) {
    return {
      backgroundColor: theme.tones.NEUTRAL.light,
      color: theme.tones.NEUTRAL.dark
    };
  }
};
function NumberBadge(_ref) {
  var _ref$tone = _ref.tone,
      tone = _ref$tone === void 0 ? "NEUTRAL" : _ref$tone,
      rest = _objectWithoutProperties(_ref, ["tone"]);

  return jsx("span", _extends({
    css: function css(theme) {
      return [baseCss(theme), toneCss[tone](theme)];
    }
  }, rest));
}
NumberBadge.propTypes = {
  tone: _pt.oneOf(["DANGER", "WARNING", "NEUTRAL"])
};
NumberBadge.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "NumberBadge",
  "props": {
    "tone": {
      "defaultValue": {
        "value": "`NEUTRAL`",
        "computed": false
      },
      "required": false,
      "tsType": {
        "name": "union",
        "raw": "`DANGER` | `WARNING` | `NEUTRAL`",
        "elements": [{
          "name": "literal",
          "value": "undefined"
        }, {
          "name": "literal",
          "value": "undefined"
        }, {
          "name": "literal",
          "value": "undefined"
        }]
      },
      "description": ""
    }
  }
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/NumberBadge/NumberBadge.tsx"] = {
    name: "NumberBadge",
    docgenInfo: NumberBadge.__docgenInfo,
    path: "src/components/NumberBadge/NumberBadge.tsx"
  };
}

export { NumberBadge };
