// gatsby-interface/dist/src/components/Tabs/Tabs.esm.js
import { extends as _extends } from '../../../_virtual/_rollupPluginBabelHelpers.js';
import { jsx } from '@emotion/core';
import React__default from 'react';
import { DisableReachStyleCheck } from '../../utils/helpers/DisableReachStyleCheck.esm.js';
import { Tabs as Tabs$1, TabsOrientation, TabList as TabList$1, TabPanels as TabPanels$1, TabPanel as TabPanel$1, Tab as Tab$1, useTabsContext as useTabsContext$1 } from '../../../node_modules/@reach/tabs/dist/tabs.esm.esm.js';
export { TabsKeyboardActivation } from '../../../node_modules/@reach/tabs/dist/tabs.esm.esm.js';
import { tabsStyles } from './Tabs.styles.esm.js';

function Tabs(props) {
  return jsx(React__default.Fragment, null, jsx(Tabs$1, _extends({
    css: tabsStyles.tabs
  }, props, {
    orientation: TabsOrientation.Horizontal
  })), jsx(DisableReachStyleCheck, {
    reachComponent: "tabs"
  }));
}
function TabList(props) {
  return jsx(TabList$1, _extends({
    css: tabsStyles.tabList
  }, props));
}
function TabPanels(props) {
  return jsx(TabPanels$1, props);
}
function TabPanel(props) {
  return jsx(TabPanel$1, props);
}
function Tab(props) {
  return jsx(Tab$1, _extends({
    css: function css(theme) {
      return [tabsStyles.tab(theme), tabsStyles.tabButton(theme), {
        "&[data-selected]": tabsStyles.tabButtonSelected(theme)
      }];
    }
  }, props));
}
function useTabsContext() {
  return useTabsContext$1();
}
Tabs.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "Tabs"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/Tabs/Tabs.tsx"] = {
    name: "Tabs",
    docgenInfo: Tabs.__docgenInfo,
    path: "src/components/Tabs/Tabs.tsx"
  };
}

TabList.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "TabList"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/Tabs/Tabs.tsx"] = {
    name: "TabList",
    docgenInfo: TabList.__docgenInfo,
    path: "src/components/Tabs/Tabs.tsx"
  };
}

TabPanels.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "TabPanels"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/Tabs/Tabs.tsx"] = {
    name: "TabPanels",
    docgenInfo: TabPanels.__docgenInfo,
    path: "src/components/Tabs/Tabs.tsx"
  };
}

TabPanel.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "TabPanel"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/Tabs/Tabs.tsx"] = {
    name: "TabPanel",
    docgenInfo: TabPanel.__docgenInfo,
    path: "src/components/Tabs/Tabs.tsx"
  };
}

Tab.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "Tab"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/Tabs/Tabs.tsx"] = {
    name: "Tab",
    docgenInfo: Tab.__docgenInfo,
    path: "src/components/Tabs/Tabs.tsx"
  };
}

export { Tab, TabList, TabPanel, TabPanels, Tabs, useTabsContext };
