// gatsby-interface/dist/src/components/Tabs/Tabs.styles.esm.js
var reachUIDefaultStyles = {
  tabs: function tabs(_theme) {
    return {
      '[data-orientation="vertical"]': {
        display: "flex"
      }
    };
  },
  tabList: function tabList(_theme) {
    return {
      display: "flex",
      '[aria-orientation="vertical"]': {
        flexDirection: "column"
      }
    };
  },
  tab: function tab(_theme) {
    return {
      display: "inline-block",
      border: "none",
      background: "none",
      color: "inherit",
      font: "inherit",
      cursor: "pointer",
      WebkitAppearance: "none",
      MozAppearance: "none"
    };
  }
};

var tabsCss = function tabsCss(theme) {
  return [reachUIDefaultStyles.tabs(theme)];
};

var tabsListCss = function tabsListCss(theme) {
  return [reachUIDefaultStyles.tabList(theme), {
    display: "flex",
    flexGrow: 1,
    flexShrink: 0,
    listStyleType: "none",
    margin: 0,
    position: "relative",
    overflowX: "auto",
    WebkitOverflowScrolling: "touch",
    "&::-webkit-scrollbar": {
      display: "none"
    },
    borderBottom: "1px solid ".concat(theme.colors.blackFade[10])
  }];
};

var tabCss = function tabCss(theme) {
  return [reachUIDefaultStyles.tab(theme), {
    flexShrink: 0,
    margin: 0,
    padding: 0,
    position: "relative",
    "&:not(:last-of-type)": {
      marginRight: theme.space[5]
    }
  }];
};

var tabButtonCss = function tabButtonCss(theme) {
  return {
    alignItems: "center",
    display: "flex",
    margin: 0,
    padding: "".concat(theme.space[4], " ").concat(theme.space[3]),
    position: "relative",
    overflowY: "hidden",
    color: theme.colors.grey[60],
    fontFamily: theme.fonts.system,
    fontSize: theme.fontSizes[2],
    background: "none",
    border: "none",
    cursor: "pointer",
    "&:after": {
      borderTopLeftRadius: theme.radii[1],
      borderTopRightRadius: theme.radii[1],
      bottom: 0,
      content: "\"\"",
      left: 0,
      height: "4px",
      position: "absolute",
      transition: "transform ".concat(theme.transitions.speed["default"]),
      width: "100%",
      background: theme.colors.purple[30],
      transform: "translateY(100%)"
    },
    "&:hover": {
      "&:after": {
        transform: "translateY(40%)"
      }
    }
  };
};

var tabButtonSelectedCss = function tabButtonSelectedCss(theme) {
  return {
    fontWeight: theme.fontWeights.bold,
    color: theme.colors.grey[70],
    "&:after": {
      background: theme.colors.purple[60],
      transform: "translateY(40%)"
    }
  };
};

var tabsStyles = {
  tabs: tabsCss,
  tabList: tabsListCss,
  tab: tabCss,
  tabButton: tabButtonCss,
  tabButtonSelected: tabButtonSelectedCss
};

export { tabsStyles };
