// gatsby-interface/dist/src/components/Announcement/Announcement.esm.js
import { extends as _extends, defineProperty as _defineProperty } from '../../../_virtual/_rollupPluginBabelHelpers.js';
import { jsx } from '@emotion/core';

var baseCss = function baseCss(theme) {
  return _defineProperty({
    display: "flex",
    alignItems: "center",
    background: theme.colors.teal[5],
    color: theme.colors.blackFade[70],
    fontSize: theme.fontSizes[1],
    fontFamily: theme.fonts.system
  }, "&:not(:first-child)", {
    borderTop: "1px solid ".concat(theme.colors.teal[10])
  });
};

var contentCss = function contentCss(theme) {
  return {
    paddingTop: theme.space[5],
    paddingBottom: theme.space[5],
    paddingLeft: theme.space[9],
    paddingRight: 0
  };
};

function Announcement(props) {
  return jsx("div", _extends({
    css: baseCss
  }, props), jsx("div", {
    css: contentCss
  }, props.children), jsx(AnnouncementDecoration, null));
}

function AnnouncementDecoration() {
  return jsx("svg", {
    width: "104",
    height: "60",
    "view-box": "0 0 104 60",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg"
  }, jsx("g", {
    opacity: "0.75"
  }, jsx("circle", {
    opacity: "0.5",
    cx: "47.4997",
    cy: "29.5",
    r: "5.50048",
    fill: "#FFDF37"
  }), jsx("ellipse", {
    opacity: "0.25",
    cx: "61.5005",
    cy: "15.5002",
    rx: "5.50048",
    ry: "5.50048",
    fill: "#BC027F"
  }), jsx("circle", {
    opacity: "0.15",
    cx: "5.49991",
    cy: "15.4997",
    r: "5.49991",
    fill: "#159BF3"
  }), jsx("ellipse", {
    opacity: "0.15",
    cx: "47.4992",
    cy: "1.49991",
    rx: "5.49991",
    ry: "5.49991",
    fill: "#B17ACC"
  }), jsx("ellipse", {
    opacity: "0.5",
    cx: "61.4995",
    cy: "57.5005",
    rx: "5.50048",
    ry: "5.50048",
    fill: "#59C156"
  }), jsx("ellipse", {
    opacity: "0.25",
    cx: "47.4997",
    cy: "43.4997",
    rx: "5.50048",
    ry: "5.50048",
    fill: "#FB8400"
  }), jsx("circle", {
    opacity: "0.15",
    cx: "61.4995",
    cy: "43.4997",
    r: "5.50048",
    fill: "#663399"
  }), jsx("ellipse", {
    opacity: "0.25",
    cx: "19.4997",
    cy: "43.4992",
    rx: "5.49991",
    ry: "5.4999",
    fill: "#05F7F4"
  }), jsx("circle", {
    opacity: "0.25",
    cx: "34",
    cy: "29.5005",
    r: "5.50048",
    fill: "#FFD280"
  }), jsx("ellipse", {
    opacity: "0.15",
    rx: "5.49991",
    ry: "5.49991",
    transform: "matrix(1 -8.74227e-08 -8.74228e-08 -1 75.4999 57.5003)",
    fill: "#159BF3"
  }), jsx("ellipse", {
    opacity: "0.15",
    rx: "5.50048",
    ry: "5.50048",
    transform: "matrix(1 -8.74229e-08 -8.74227e-08 -1 103.5 29.5002)",
    fill: "#A1DA9E"
  }), jsx("ellipse", {
    opacity: "0.25",
    rx: "5.4999",
    ry: "5.49991",
    transform: "matrix(1 -8.74229e-08 -8.74227e-08 -1 89.4997 29.5008)",
    fill: "#159BF3"
  }), jsx("ellipse", {
    opacity: "0.25",
    rx: "5.50048",
    ry: "5.50048",
    transform: "matrix(1 -8.74229e-08 -8.74227e-08 -1 103.5 43.5)",
    fill: "#FB8400"
  })));
}

Announcement.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "Announcement"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/Announcement/Announcement.tsx"] = {
    name: "Announcement",
    docgenInfo: Announcement.__docgenInfo,
    path: "src/components/Announcement/Announcement.tsx"
  };
}

export { Announcement };
