// gatsby-interface/dist/src/components/BuildLogs/BuildLogEntrySkeleton.esm.js
import { objectWithoutProperties as _objectWithoutProperties, extends as _extends, defineProperty as _defineProperty } from '../../../_virtual/_rollupPluginBabelHelpers.js';
import { jsx } from '@emotion/core';
import _pt from 'prop-types';

var baseCss = function baseCss(theme) {
  return _defineProperty({
    fontSize: theme.fontSizes[1],
    wordBreak: "break-word",
    padding: "".concat(theme.space[4], " 0"),
    position: "relative",
    display: "grid",
    alignItems: "center",
    gridTemplateColumns: "30px auto minmax(auto, 125px)"
  }, theme.mediaQueries.tablet, {
    gridTemplateColumns: "30px minmax(auto, 50%) 120px minmax(100px, 28%)"
  });
};

var iconContainerCss = function iconContainerCss(theme) {
  return {
    alignSelf: "start",
    display: "flex",
    alignItems: "center",
    paddingTop: theme.space[1]
  };
};

function BuildLogEntrySkeleton(_ref2) {
  var icon = _ref2.icon,
      children = _ref2.children,
      rest = _objectWithoutProperties(_ref2, ["icon", "children"]);

  return jsx("div", _extends({
    css: baseCss
  }, rest), jsx("div", {
    css: iconContainerCss
  }, icon), children);
}
BuildLogEntrySkeleton.propTypes = {
  icon: _pt.node.isRequired
};
BuildLogEntrySkeleton.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "BuildLogEntrySkeleton",
  "props": {
    "icon": {
      "required": true,
      "tsType": {
        "name": "ReactReactNode",
        "raw": "React.ReactNode"
      },
      "description": ""
    }
  }
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/BuildLogs/BuildLogEntrySkeleton.tsx"] = {
    name: "BuildLogEntrySkeleton",
    docgenInfo: BuildLogEntrySkeleton.__docgenInfo,
    path: "src/components/BuildLogs/BuildLogEntrySkeleton.tsx"
  };
}

export { BuildLogEntrySkeleton };
