// gatsby-interface/dist/src/components/BuildLogs/BuildLogsList.esm.js
import { extends as _extends } from '../../../_virtual/_rollupPluginBabelHelpers.js';
import { jsx } from '@emotion/core';
import { BuildStandardLogEntry } from './BuildStandardLogEntry.esm.js';
import { BuildActivityEntry } from './BuildActivityEntry.esm.js';

var baseCss = function baseCss(_theme) {
  return {
    listStyle: "none",
    margin: 0,
    padding: 0
  };
};

function BuildLogsList(_ref) {
  var logItems = _ref.logItems,
      ariaLabel = _ref["aria-label"],
      className = _ref.className;
  return jsx("div", {
    role: "log",
    className: className
  }, jsx("ol", {
    css: baseCss,
    "aria-label": ariaLabel
  }, logItems.map(function (logItem) {
    return jsx("li", {
      css: {
        margin: 0
      },
      key: logItem.id
    }, logItem.activity ? jsx(BuildActivityEntry, _extends({}, logItem, {
      activity: logItem.activity
    })) : jsx(BuildStandardLogEntry, {
      level: logItem.level,
      message: logItem.message,
      context: logItem.context,
      filePath: logItem.filePath
    }));
  })));
}
BuildLogsList.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "BuildLogsList",
  "props": {
    "logItems": {
      "required": true,
      "tsType": {
        "name": "Array",
        "elements": [{
          "name": "BuildLogItem"
        }],
        "raw": "BuildLogItem[]"
      },
      "description": ""
    },
    "aria-label": {
      "required": true,
      "tsType": {
        "name": "string"
      },
      "description": ""
    },
    "className": {
      "required": false,
      "tsType": {
        "name": "string"
      },
      "description": ""
    }
  }
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/BuildLogs/BuildLogsList.tsx"] = {
    name: "BuildLogsList",
    docgenInfo: BuildLogsList.__docgenInfo,
    path: "src/components/BuildLogs/BuildLogsList.tsx"
  };
}

export { BuildLogsList };
