// gatsby-interface/dist/src/components/BuildLogs/types.esm.js
var BuildActivityStatus;

(function (BuildActivityStatus) {
  BuildActivityStatus["InProgress"] = "IN_PROGRESS";
  BuildActivityStatus["NotStarted"] = "NOT_STARTED";
  BuildActivityStatus["Failed"] = "FAILED";
  BuildActivityStatus["Interrupted"] = "INTERRUPTED";
  BuildActivityStatus["Success"] = "SUCCESS";
})(BuildActivityStatus || (BuildActivityStatus = {}));

var BuildActivityType;

(function (BuildActivityType) {
  BuildActivityType["Progress"] = "PROGRESS";
  BuildActivityType["Spinner"] = "SPINNER";
  BuildActivityType["Pending"] = "PENDING";
  BuildActivityType["Success"] = "SUCCESS";
  BuildActivityType["Hidden"] = "HIDDEN";
})(BuildActivityType || (BuildActivityType = {}));

var StructuredLogLevel;

(function (StructuredLogLevel) {
  StructuredLogLevel["All"] = "ALL";
  StructuredLogLevel["Log"] = "LOG";
  StructuredLogLevel["Warning"] = "WARNING";
  StructuredLogLevel["Info"] = "INFO";
  StructuredLogLevel["Success"] = "SUCCESS";
  StructuredLogLevel["Debug"] = "DEBUG";
  StructuredLogLevel["Error"] = "ERROR";
})(StructuredLogLevel || (StructuredLogLevel = {}));

export { BuildActivityStatus, BuildActivityType, StructuredLogLevel };
