// @apollo/space-kit/MenuHeading/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var _tslib = require('../_tslib-bcbe0269.js');
var core = require('@emotion/core');
var React = require('react');
var React__default = _interopDefault(React);
var colors = require('../colors/colors.js');
require('../assertUnreachable-18645656.js');
require('tinycolor2');
require('../index-272aedc8.js');
var index = require('../MenuItem/index.js');

/**
 * Heading intended to be used in a menu. Can also be used to deliniate between
 * sections in a single menu.
 *
 * Composed with and accepts all props of `MenuItem`
 */
var MenuHeading = React__default.forwardRef(function (_a, ref) {
    var children = _a.children, count = _a.count, props = _tslib.__rest(_a, ["children", "count"]);
    return (core.jsx(index.MenuItem, _tslib.__assign({}, props, { interactive: false, ref: ref }),
        core.jsx("h2", { css: core.css({
                display: "flex",
                fontSize: "inherit",
                fontWeight: 600,
                margin: 0,
                marginBottom: 2,
                marginTop: 3,
                textTransform: "uppercase",
                width: "100%",
            }), ref: ref },
            core.jsx("span", { css: core.css({ flex: "1" }) }, children),
            count && (core.jsx("span", { css: core.css({
                    color: colors.colors.grey.light,
                    flex: "none",
                    fontWeight: "normal",
                }) }, count)))));
});

exports.MenuHeading = MenuHeading;
