// gatsby-interface/dist/src/components/AnchorButton/AnchorButton.esm.js
import { objectWithoutProperties as _objectWithoutProperties, extends as _extends } from '../../../_virtual/_rollupPluginBabelHelpers.js';
import { jsx } from '@emotion/core';
import React__default from 'react';
import { BaseAnchor } from '../BaseAnchor/BaseAnchor.esm.js';
import { getButtonStyles } from '../Button/Button.esm.js';

var AnchorButton = React__default.forwardRef(function (props, ref) {
  var children = props.children,
      size = props.size,
      tone = props.tone,
      variant = props.variant,
      leftIcon = props.leftIcon,
      rightIcon = props.rightIcon,
      textVariant = props.textVariant,
      width = props.width,
      rest = _objectWithoutProperties(props, ["children", "size", "tone", "variant", "leftIcon", "rightIcon", "textVariant", "width"]);

  return jsx(BaseAnchor, _extends({}, getButtonStyles({
    children: children,
    size: size,
    tone: tone,
    variant: variant,
    leftIcon: leftIcon,
    rightIcon: rightIcon,
    textVariant: textVariant,
    width: width
  }), rest, {
    ref: ref
  }));
});
AnchorButton.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "AnchorButton"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/AnchorButton/AnchorButton.tsx"] = {
    name: "AnchorButton",
    docgenInfo: AnchorButton.__docgenInfo,
    path: "src/components/AnchorButton/AnchorButton.tsx"
  };
}

export { AnchorButton };
