// gatsby-interface/dist/src/components/Toggle/ToggleGutter.esm.js
import { extends as _extends } from '../../../_virtual/_rollupPluginBabelHelpers.js';
import { jsx } from '@emotion/core';

function ToggleGutter(props) {
  return jsx("span", _extends({}, props, {
    "aria-hidden": true,
    css: function css(theme) {
      return [{
        background: theme.colors.grey[30],
        borderRadius: theme.radii[5],
        cursor: "pointer",
        display: "inline-block",
        height: "24px",
        padding: "3px",
        transition: "all .3s ease, background .5s",
        userSelect: "none",
        width: "48px",
        ":after": {
          background: theme.colors.white,
          position: "relative",
          display: "block",
          content: "\"\"",
          width: "18px",
          height: "18px",
          borderRadius: theme.radii[5],
          transition: "all 0.1s ease",
          left: 0
        },
        ":active": {
          ":after": {
            width: "24px"
          }
        }
      }];
    }
  }));
}
var ToggleGutterTagName = "span";
var toggleGutterFocusCss = function toggleGutterFocusCss(theme) {
  return {
    boxShadow: "0 0 0 3px ".concat(theme.colors.blue[30]),
    outline: "0"
  };
};
var toggleGutterCheckedCss = function toggleGutterCheckedCss() {
  var tone = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : "BRAND";
  return function (theme) {
    return {
      background: theme.tones[tone].medium,
      ":after": {
        left: "calc(100% - 18px)"
      },
      ":active": {
        ":after": {
          left: "calc(100% - 24px)"
        }
      }
    };
  };
};
ToggleGutter.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "ToggleGutter"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/Toggle/ToggleGutter.tsx"] = {
    name: "ToggleGutter",
    docgenInfo: ToggleGutter.__docgenInfo,
    path: "src/components/Toggle/ToggleGutter.tsx"
  };
}

export default ToggleGutter;
export { ToggleGutterTagName, toggleGutterCheckedCss, toggleGutterFocusCss };
