// gatsby-interface/dist/src/components/LinkButton/LinkButton.esm.js
import { objectWithoutProperties as _objectWithoutProperties, extends as _extends } from '../../../_virtual/_rollupPluginBabelHelpers.js';
import { jsx } from '@emotion/core';
import { getButtonStyles } from '../Button/Button.esm.js';
import { BaseLink } from '../BaseLink/BaseLink.esm.js';

function LinkButton(props) {
  var children = props.children,
      size = props.size,
      tone = props.tone,
      variant = props.variant,
      leftIcon = props.leftIcon,
      rightIcon = props.rightIcon,
      textVariant = props.textVariant,
      width = props.width,
      rest = _objectWithoutProperties(props, ["children", "size", "tone", "variant", "leftIcon", "rightIcon", "textVariant", "width"]);

  return jsx(BaseLink, _extends({}, getButtonStyles({
    children: children,
    size: size,
    tone: tone,
    variant: variant,
    leftIcon: leftIcon,
    rightIcon: rightIcon,
    textVariant: textVariant,
    width: width
  }), rest));
}
LinkButton.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "LinkButton"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/LinkButton/LinkButton.tsx"] = {
    name: "LinkButton",
    docgenInfo: LinkButton.__docgenInfo,
    path: "src/components/LinkButton/LinkButton.tsx"
  };
}

export { LinkButton };
