// gatsby-interface/dist/src/components/BaseAnchor/BaseAnchor.esm.js
import { objectWithoutProperties as _objectWithoutProperties, extends as _extends } from '../../../_virtual/_rollupPluginBabelHelpers.js';
import React__default from 'react';
import secureTargetBlankLink from '../../utils/helpers/secureTargetBlankLink.esm.js';

var BaseAnchor = React__default.forwardRef(function (_ref, ref) {
  var role = _ref.role,
      _ref$target = _ref.target,
      target = _ref$target === void 0 ? "_blank" : _ref$target,
      rel = _ref.rel,
      children = _ref.children,
      rest = _objectWithoutProperties(_ref, ["role", "target", "rel", "children"]);

  return /*#__PURE__*/React__default.createElement("a", _extends({
    target: target,
    role: role,
    rel: secureTargetBlankLink({
      rel: rel,
      target: target
    })
  }, rest, {
    ref: ref
  }), children);
});
BaseAnchor.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "BaseAnchor",
  "props": {
    "target": {
      "defaultValue": {
        "value": "`_blank`",
        "computed": false
      },
      "required": false
    }
  }
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/BaseAnchor/BaseAnchor.tsx"] = {
    name: "BaseAnchor",
    docgenInfo: BaseAnchor.__docgenInfo,
    path: "src/components/BaseAnchor/BaseAnchor.tsx"
  };
}

export { BaseAnchor };
