// gatsby-interface/dist/src/components/Modal/ModalCard.esm.js
import { extends as _extends, taggedTemplateLiteral as _taggedTemplateLiteral, defineProperty as _defineProperty } from '../../../_virtual/_rollupPluginBabelHelpers.js';
import { keyframes, jsx } from '@emotion/core';
import { ModalContent } from './ModalContent.esm.js';

function _templateObject() {
  var data = _taggedTemplateLiteral(["\n  100% {\n     transform: translate(0, calc(50vh - 50%)) scale(1) perspective(1000px) rotateX(0);\n  }\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}
var cardIncoming = keyframes(_templateObject());

var baseCss = function baseCss(theme) {
  var _ref;

  return _ref = {
    background: theme.colors.white,
    borderRadius: theme.radii[3],
    boxShadow: "0px 5px 30px rgba(0, 0, 0, 0.2)",
    maxWidth: "calc(100% - (".concat(theme.space[5], " * 2))"),
    animation: "".concat(cardIncoming, " 0.5s 0.25s ease forwards"),
    transform: "translate(0, 90vh) scale(0.9) perspective(1000px) rotateX(-90deg)",
    transformOrigin: "top center",
    margin: "0 auto"
  }, _defineProperty(_ref, theme.mediaQueries.mobile, {
    maxWidth: "calc(100% - (".concat(theme.space[7], " * 2))")
  }), _defineProperty(_ref, theme.mediaQueries.phablet, {
    width: "620px"
  }), _ref;
};

var ModalCard = function ModalCard(props) {
  return jsx(ModalContent, _extends({
    css: baseCss
  }, props));
};
ModalCard.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "ModalCard"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/Modal/ModalCard.tsx"] = {
    name: "ModalCard",
    docgenInfo: ModalCard.__docgenInfo,
    path: "src/components/Modal/ModalCard.tsx"
  };
}

export { ModalCard };
