diff --git a/node_modules/@ethersproject/providers/lib/formatter.js b/node_modules/@ethersproject/providers/lib/formatter.js
index 6f5fa9d..e271f08 100644
--- a/node_modules/@ethersproject/providers/lib/formatter.js
+++ b/node_modules/@ethersproject/providers/lib/formatter.js
@@ -266,6 +266,11 @@ var Formatter = /** @class */ (function () {
         if (transaction.gas != null && transaction.gasLimit == null) {
             transaction.gasLimit = transaction.gas;
         }
+
+        if (transaction.to && transaction.to.length === 43 && transaction.to.slice(0, 3) === 'xdc') {
+          transaction.to = '0x' + transaction.to.slice(3);
+        }
+
         // Some clients (TestRPC) do strange things like return 0x0 for the
         // 0 address; correct this to be a real address
         if (transaction.to && bignumber_1.BigNumber.from(transaction.to).isZero()) {
