// gatsby-interface/dist/src/components/Badge/Badge.helpers.esm.js
var variants = {
  STATUS: function STATUS(tone) {
    return function (theme) {
      return [{
        background: theme.tones[tone].superLight,
        border: "1px solid ".concat(theme.tones[tone].lighter),
        color: theme.tones[tone].text
      }];
    };
  },
  PILL: function PILL(tone) {
    return function (theme) {
      return [{
        background: theme.tones[tone].dark,
        color: theme.tones[tone].textInverted ? theme.tones[tone].textInverted : theme.colors.white
      }];
    };
  }
};
function getBadgeVariantStyles(variant, tone) {
  return variants[variant](tone);
}
getBadgeVariantStyles.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "getBadgeVariantStyles"
};

if (typeof STORYBOOK_REACT_CLASSES !== "undefined") {
  STORYBOOK_REACT_CLASSES["src/components/Badge/Badge.helpers.tsx"] = {
    name: "getBadgeVariantStyles",
    docgenInfo: getBadgeVariantStyles.__docgenInfo,
    path: "src/components/Badge/Badge.helpers.tsx"
  };
}

export { getBadgeVariantStyles };
