// contentful-management/dist/es-modules/entities/environment.js
import cloneDeep from 'lodash/cloneDeep';
import { freezeSys, toPlainObject } from 'contentful-sdk-core';
import enhanceWithMethods from '../enhance-with-methods';
import createEnvironmentApi from '../create-environment-api';
import { wrapCollection } from '../common-utils';

/**
 * This method creates the API for the given environment with all the methods for
 * reading and creating other entities. It also passes down a clone of the
 * http client with a environment id, so the base path for requests now has the
 * environment id already set.
 * @private
 * @param http - HTTP client instance
 * @param data - API response for a Environment
 * @return
 */
export function wrapEnvironment(http, data) {
  // do not pollute generated typings
  var sdkHttp = http;
  var environment = toPlainObject(cloneDeep(data));
  var _sdkHttp$httpClientPa = sdkHttp.httpClientParams,
      hostUpload = _sdkHttp$httpClientPa.hostUpload,
      defaultHostnameUpload = _sdkHttp$httpClientPa.defaultHostnameUpload;
  var environmentScopedHttpClient = sdkHttp.cloneWithNewParams({
    baseURL: http.defaults.baseURL + 'environments/' + environment.sys.id
  });
  var environmentScopedUploadClient = sdkHttp.cloneWithNewParams({
    space: environment.sys.space.sys.id,
    host: hostUpload || defaultHostnameUpload
  });
  var environmentApi = createEnvironmentApi({
    http: environmentScopedHttpClient,
    httpUpload: environmentScopedUploadClient
  });
  var enhancedEnvironment = enhanceWithMethods(environment, environmentApi);
  return freezeSys(enhancedEnvironment);
}
/**
 * This method wraps each environment in a collection with the environment API. See wrapEnvironment
 * above for more details.
 * @private
 */

export var wrapEnvironmentCollection = wrapCollection(wrapEnvironment);
