// contentful-management/dist/es-modules/entities/environment-alias.js
import cloneDeep from 'lodash/cloneDeep';
import { freezeSys, toPlainObject } from 'contentful-sdk-core';
import enhanceWithMethods from '../enhance-with-methods';
import { createUpdateEntity } from '../instance-actions';
import { wrapCollection } from '../common-utils';

function createEnvironmentAliasApi(http) {
  return {
    update: createUpdateEntity({
      http: http,
      entityPath: 'environment_aliases',
      wrapperMethod: wrapEnvironmentAlias
    })
  };
}
/**
 * @private
 * @param http - HTTP client instance
 * @param data - Raw environment alias data
 * @return Wrapped environment alias data
 */


export function wrapEnvironmentAlias(http, data) {
  var alias = toPlainObject(cloneDeep(data));
  var enhancedAlias = enhanceWithMethods(alias, createEnvironmentAliasApi(http));
  return freezeSys(enhancedAlias);
}
/**
 * @private
 * @param http - HTTP client instance
 * @param data - Raw environment alias collection data
 * @return Wrapped environment alias collection data
 */

export var wrapEnvironmentAliasCollection = wrapCollection(wrapEnvironmentAlias);
