// contentful-management/dist/es-modules/entities/index.js
import * as appDefinition from './app-definition';
import * as space from './space';
import * as environment from './environment';
import * as entry from './entry';
import * as asset from './asset';
import * as contentType from './content-type';
import * as editorInterface from './editor-interface';
import * as locale from './locale';
import * as webhook from './webhook';
import * as spaceMember from './space-member';
import * as spaceMembership from './space-membership';
import * as teamSpaceMembership from './team-space-membership';
import * as organizationMembership from './organization-membership';
import * as organizationInvitation from './organization-invitation';
import * as role from './role';
import * as apiKey from './api-key';
import * as previewApiKey from './preview-api-key';
import * as upload from './upload';
import * as organization from './organization';
import * as uiExtension from './ui-extension';
import * as appInstallation from './app-installation';
import * as snapshot from './snapshot';
import * as user from './user';
import * as personalAccessToken from './personal-access-token';
import * as usage from './usage';
import * as environmentAlias from './environment-alias';
import * as team from './team';
import * as teamMembership from './team-membership';
import * as tag from './tag';
export default {
  appDefinition: appDefinition,
  space: space,
  environment: environment,
  entry: entry,
  asset: asset,
  contentType: contentType,
  editorInterface: editorInterface,
  locale: locale,
  webhook: webhook,
  spaceMember: spaceMember,
  spaceMembership: spaceMembership,
  teamSpaceMembership: teamSpaceMembership,
  organizationMembership: organizationMembership,
  organizationInvitation: organizationInvitation,
  role: role,
  apiKey: apiKey,
  previewApiKey: previewApiKey,
  upload: upload,
  organization: organization,
  uiExtension: uiExtension,
  appInstallation: appInstallation,
  snapshot: snapshot,
  user: user,
  personalAccessToken: personalAccessToken,
  usage: usage,
  environmentAlias: environmentAlias,
  team: team,
  teamMembership: teamMembership,
  tag: tag
};
