// contentful-management/dist/es-modules/entities/ui-extension.js
import cloneDeep from 'lodash/cloneDeep';
import { freezeSys, toPlainObject } from 'contentful-sdk-core';
import enhanceWithMethods from '../enhance-with-methods';
import { createUpdateEntity, createDeleteEntity } from '../instance-actions';
import { wrapCollection } from '../common-utils';

function createUiExtensionApi(http) {
  return {
    update: createUpdateEntity({
      http: http,
      entityPath: 'extensions',
      wrapperMethod: wrapUiExtension
    }),
    "delete": createDeleteEntity({
      http: http,
      entityPath: 'extensions'
    })
  };
}
/**
 * @private
 * @param http - HTTP client instance
 * @param data - Raw UI Extension data
 * @return Wrapped UI Extension data
 */


export function wrapUiExtension(http, data) {
  var uiExtension = toPlainObject(cloneDeep(data));
  var uiExtensionWithMethods = enhanceWithMethods(uiExtension, createUiExtensionApi(http));
  return freezeSys(uiExtensionWithMethods);
}
/**
 * @private
 */

export var wrapUiExtensionCollection = wrapCollection(wrapUiExtension);
